crates/red_knot/src/program/check.rs

```diff
@@ -281,30 +253,35 @@ impl<'a> CheckFilesLoop<'a> {
                     }
                 }
                 CheckFileMessage::Queue(id) => {
-                    self.queue_file(id, context.clone())?;
+                    if !cancelled {
+                        self.queue_file(id, context.clone());
+                    }
                 }
                 CheckFileMessage::Cancelled => {
-                    return Err(CheckError::Cancelled);
+                    self.pending -= 1;
+                    cancelled = true;
+
+                    if self.pending == 0 {
+                        break;
+                    }
                 }
             }
         }
 
-        Ok(result)
-    }
-
-    fn queue_file(&mut self, file_id: FileId, context: CheckContext) -> Result<(), CheckError> {
-        if context.is_cancelled() {
-            return Err(CheckError::Cancelled);
+        if cancelled {
+            Err(QueryError::Cancelled)
+        } else {
+            Ok(result)
```

Comment:
Shouldn't we return `Err(QueryError::Cancelled)` immediately if we get `CheckFileMessage::Cancelled`, instead of waiting for the loop to finish?

Reply:
The problem with returning immediately is that we then drop the only reference to `receiver`. That means, that any `sender.send(message).unwrap()` calls will fail in the threads checking the files. I'm not sure if there's a more idiomatic way of doing this other than waiting for all file check operations to complete to be sure there will be no more incoming messages and only then exit.

Reply:
Can't we just handling that error in the threads checking the files though? Like, if `sender.send(message)` fails, we can still exit the thread gracefully instead of panicking.

Reply:
I guess we could, but it would mean that it will be harder to find the bug if we drop the receiver accidentally for another reason. Anyway. The next PR reworked this quiet significantly and I think there we can return immediately.